src/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java

```diff
@@ -67,7 +67,14 @@
     private final Map<String, RecordedConsumer> consumers = new ConcurrentHashMap<String, RecordedConsumer>();
     private final List<ConsumerRecoveryListener> consumerRecoveryListeners = new ArrayList<ConsumerRecoveryListener>();
     private final List<QueueRecoveryListener> queueRecoveryListeners = new ArrayList<QueueRecoveryListener>();
-
+	
+	//Used to block connection recovery attempts after close() is invoked.
+	private volatile boolean manuallyClosed = false;
+	
+	//This lock guards the manuallyClosed flag and the delegate connection.  Guarding these two ensures that a new connection can never
+	//be created after application code has initiated shutdown.  
+	private Object recoveryLock = new Object();
```

Comment:
beginAutomaticRecovery() was marked synchronized, and holds that lock
throughout the connection recovery process.  I would not be able to
synchronize on the AutorecoveringConnection object from within the
close/abort methods without locking up the application code.  That's why I
created the secondary lock object.

On Tue, Jul 21, 2015 at 5:59 PM, Michael Klishin <notifications@github.com>
wrote:

> In src/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java
> <https://github.com/rabbitmq/rabbitmq-java-client/pull/76#discussion_r35159924>
> :
>
> > @@ -67,7 +67,14 @@
> >      private final Map<String, RecordedConsumer> consumers = new ConcurrentHashMap<String, RecordedConsumer>();
> >      private final List<ConsumerRecoveryListener> consumerRecoveryListeners = new ArrayList<ConsumerRecoveryListener>();
> >      private final List<QueueRecoveryListener> queueRecoveryListeners = new ArrayList<QueueRecoveryListener>();
> > -
> > +	
> > +	//Used to block connection recovery attempts after close() is invoked.
> > +	private volatile boolean manuallyClosed = false;
> > +	
> > +	//This lock guards the manuallyClosed flag and the delegate connection.  Guarding these two ensures that a new connection can never
> > +	//be created after application code has initiated shutdown.
> > +	private Object recoveryLock = new Object();
>
> Synchronising on this is not discouraged in Java.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-java-client/pull/76/files#r35159924>
> .
>



-- 
 *  Daniel Hakim*
  TRX Systems Inc.
  dhakim@trxsystems.com
  Phone: +1 (301) 313-0053 Ext 731

Developer of NEON™ Personnel Location Products
www.trxsystems.com

 The information contained in this e-mail and any attachment is covered by
the Electronic Communications Privacy Act, 18 U.S.C. 2510-2521, may contain
TRX Systems, Inc. proprietary and/or confidential information, and is
intended only for the confidential use of the above named recipient. Please
contact the sender if you receive this email and it is not intended for you.

Follow TRX Systems: <http://feeds.feedburner.com/trxsystems/hXJN>
<http://www.linkedin.com/companies/932098> <http://twitter.com/TRX_SYSTEMS>